remotes/docker/pusher.go

```diff
@@ -398,6 +400,7 @@ func (pw *pushWriter) Write(p []byte) (n int, err error) {
 func (pw *pushWriter) Close() error {
 	// Ensure pipeC is closed but handle `Close()` being
 	// called multiple times without panicking
+	pw.lock.Lock()
```

Comment:
The `Close()` already makes sure that multiple calls to close the channel can be handled without panicking. Then how does this occur . Is it because the channel is closed in a different goroutine after a call to 
```golang
402:     _, ok := <-pw.pipeC
403:        if ok {
```
i.e `Close()` is called in a different go routine between 402 and 403 of the current routine

Reply:
According to my debug, it is because there are two goroutine calls to the `Close()` method, which causes pain in `close(pw.pipeC)`.

Reply:
You can choose not to lock here, but you need to solve the problem of concurrent calling the `(pw *pushWriter)Close()` method.

Reply:
I think it might be cleaner to use sync.Once rather than locking, and there is no need to use select.
eg.
```go
pw.closeOnce.Do(func(){
    close(pw.pipeC)
})
```

Reply:
> I think it might be cleaner to use sync.Once rather than locking, and there is no need to use select. eg.
> 
> ```go
> pw.closeOnce.Do(func(){
>     close(pw.pipeC)
> })
> ```

That's good advice.